File: apps/desktop/src/api/endpoints/security.ts
import client from './client'

export interface EncryptOptions {
  input_path: string
  output_path: string
  user_password: string
  owner_password?: string
  permissions?: {
    print?: boolean
    copy?: boolean
    modify?: boolean
    annotate?: boolean
  }
}

export interface DecryptOptions {
  input_path: string
  output_path: string
  password: string
}

export interface SignOptions {
  input_path: string
  output_path: string
  certificate_path: string
  certificate_password: string
  reason?: string
  location?: string
}

export const securityApi = {
  async encrypt(options: EncryptOptions): Promise<{ success: boolean; output_path: string }> {
    return client.post('/security/encrypt', options)
  },

  async decrypt(options: DecryptOptions): Promise<{ success: boolean; output_path: string }> {
    return client.post('/security/decrypt', options)
  },

  async sign(options: SignOptions): Promise<{ success: boolean; output_path: string }> {
    return client.post('/security/sign', options)
  },

  async verifySignature(filePath: string): Promise<{
    valid: boolean
    signer?: string
    signed_at?: string
    reason?: string
    location?: string
  }> {
    return client.post('/security/verify', null, {
      params: { file_path: filePath }
    })
  },

  async getPermissions(
    filePath: string,
    password?: string
  ): Promise<{
    encrypted: boolean
    permissions?: {
      print: boolean
      copy: boolean
      modify: boolean
      annotate: boolean
    }
  }> {
    return client.get(`/security/permissions/${encodeURIComponent(filePath)}`, {
      params: { password }
    })
  },

  async redact(
    inputPath: string,
    outputPath: string,
    redactions: Array<{
      page: number
      x: number
      y: number
      width: number
      height: number
    }>
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/security/redact', {
      input_path: inputPath,
      output_path: outputPath,
      redactions
    })
  }
}

export default securityApi

File: apps/server/app/api/v1/security.py
from fastapi import APIRouter, HTTPException
from typing import Optional
from pydantic import BaseModel

from app.core.security_engine.engine import SecurityEngine

router = APIRouter()
security_engine = SecurityEngine()


class EncryptRequest(BaseModel):
    input_path: str
    output_path: str
    user_password: str
    owner_password: Optional[str] = None
    permissions: dict = {
        "print": True,
        "copy": False,
        "modify": False,
        "annotate": True,
    }


class DecryptRequest(BaseModel):
    input_path: str
    output_path: str
    password: str


class SignRequest(BaseModel):
    input_path: str
    output_path: str
    certificate_path: str
    certificate_password: str
    reason: Optional[str] = None
    location: Optional[str] = None


@router.post("/encrypt")
async def encrypt_pdf(request: EncryptRequest):
    try:
        result = await security_engine.encrypt(
            request.input_path,
            request.output_path,
            request.user_password,
            request.owner_password,
            request.permissions
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/decrypt")
async def decrypt_pdf(request: DecryptRequest):
    try:
        result = await security_engine.decrypt(
            request.input_path,
            request.output_path,
            request.password
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/sign")
async def sign_pdf(request: SignRequest):
    try:
        result = await security_engine.sign(
            request.input_path,
            request.output_path,
            request.certificate_path,
            request.certificate_password,
            request.reason,
            request.location
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/verify")
async def verify_signature(file_path: str):
    try:
        result = await security_engine.verify_signature(file_path)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/permissions/{file_path:path}")
async def get_permissions(file_path: str, password: Optional[str] = None):
    try:
        permissions = await security_engine.get_permissions(file_path, password)
        return permissions
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/redact")
async def redact_pdf(
    input_path: str,
    output_path: str,
    redactions: list[dict]
):
    try:
        result = await security_engine.redact(input_path, output_path, redactions)
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
